Automattic/wp-calypso
# Gutenberg Integration: load editor with a new post

This PR adds the **bare minimum** wiring to Calypso to support the API v2 Post format, so Gutenberg can somewhat start to function.

Depends on D18064-code

![screen shot 2018-09-07 at 14 21 32](https://user-images.githubusercontent.com/233601/45233571-87102700-b2a9-11e8-97d7-e0b22e29ec61.png)

Gutenberg relies heavily on the provided `post` prop, with a schema as returned by the `wp/v2` API when the `edit` context is provided. Without a valid post, Gutenberg fails to load most of the sidebar tools.
This PR adds as little code as possible to wire the `wp/v2` endpoints needed to create a draft and load the post so Gutenberg can use the returned `post`.

This PR is based on the `createNewPost` function on [`p2-gutenberg`](https://github.com/Automattic/p2-gutenberg/blob/460c72570cd073f47b2dfab4a9eb0cde6fcbc14c/p2editor/api/index.js#L1):

```js
export const createNewPost = () => {
	return new Promise( ( resolve ) => {
		jQuery.wpcom_proxy_request( {
			path: `/sites/${ window._currentSiteId }/p2/post`,
			apiNamespace: 'wpcom/v2',
		} ).then( ( { ID } ) => {
			wp.apiRequest( {
				path: `/wp/v2/posts/${ ID }?context=edit`,
			} ).then( resolve );
		} );
	} );
};
```

~~### State~~

~~The `post` stored on Calypso's redux state is insufficient for Gutenberg, that relies on fields returned by the API when querying for a post and the `edit` context is provided. In particular, the sidebar uses the `_links` metadata returned by the API to decide whether or not to populate hierarchical and flat lists like _categories_ and _taxonomies_, despite the fact that those values are queried and stored on `core` and `core/data` redux store.
Gutenberg stores the current post information on the `currentPost` key on the  `core/editor` store. Calypso has to do something similar, so, on the `gutenberg.currentPost` branch on Calypso's redux store, we keep the information returned from the API call.~~

~~### Query Components~~

~~This PR also incorporates two query components, `QueryGutenbergCreatePost` to create a new auto draft, and `QueryGutenbergSitePost` to query any post using the `wp/v2` endpoint.~~

~~The App Component holding the Gutenberg instance implements only `QueryGutenbergCreatePost`. When the page load, a new draft is created, fetched, loaded into state and sent over the `props` to Gutenberg. Refresh the page, and a new one is created.~~

~~### Data Layer~~

~~This PR also breaks new ground because it implements the first two endpoints for the `wp/v2` API namespace for calypso. I've made the controversial decision of creating a new handler (`wpHandlers`) and a new top level folder structure to keep things apart.~~

~~The `requestGutenbergPostDraft` request handler dispatches the fetch of the newly created draft on success, and there are no state management associated with that action. I initially implemented tracking on different active drafts, but removed all that complexity in favor of fixing the rest of the integration problems. We can worry about multiple instances of Gutenberg in the future.~~

### How To Test and What to Expect

Do not expect much 😎

~~There are some basic unit tests for the state management and the data layer that can be run with:~~

To test the Gutenberg integration, you'll have to apply D18064-code and sandbox `public-api.wordpress.com` to use the new create draft endpoint. Then, navigate to `/gutenberg/post/{site-slug}`.

Gutenberg should load with all the sidebar elements active. The _Categories_ panel should load all the available categories, and the authors should load on the _Status and Visibility_ panel. Taxonomies load on the store, but for reason to explore on another PR they are not loading on the token input.

The _Categories_ block should load all the categories, but _autosave_ and the rest of the functionality is sadly broken. This PR should enable work on those features.
